005-1-locations-on-stack.md

```diff
@@ -0,0 +1,590 @@
+> value of type `T`.
+> 
+> _While the size of the pushed value could be inferred from the base type
+> definition, it is encoded explicitly into the operation so that the
+> operation can be parsed easily without reference to the `.debug\_info`
+> section._
+
+Move items 13 (`DW_OP_push_object_address`), 14 (`DW_OP_form_tls_address`),
+and 15 (`DW_OP_call_frame_cfa`) to (new) Section 2.5.2 Location Operations.
+
+
+### Section 2.5.2 Location Operations [NEW]
+
+Insert the following (adapted from parts of Section 2.6) into this new section:
+
+> Information about the location of program objects is provided by
+> location descriptions. Location descriptions can be either of two forms:
+> 
+> 1. Single location descriptions, which are a language-independent
+> representation of addressing rules of arbitrary complexity built from
+> DWARF expressions. They are sufficient for describing the location of
+> any object as long as its lifetime is either static or the same as the
+> lexical block that owns it, and it does not move during its lifetime.
+> 
+> 2. Location lists, which are used to describe objects that have a
+> limited lifetime or change their location during their lifetime.
+> Location lists are described in Section 2.6.
+> 
+> Location descriptions are distinguished in a context-sensitive manner.
+> As the value of an attribute, a single location description is encoded
+> using class `locdesc`, and a location list is encoded using class `loclist`
+> (which serves as an index into a separate section containing location
+> lists).
+> 
+> The following operation can be used to push a location onto the stack:
+> 
+> 1. `DW_OP_push_object_address` [moved from Section 2.5.1.3]  
+> The DW_OP_push_object_address operation pushes the location of the
+> object currently being evaluated as part of evaluation of a user
+> presented expression. This object may correspond to an independent
+> variable described by its own debugging information entry or it may be a
+> component of an array, structure, or class whose address has been
+> dynamically determined by an earlier step during user expression
+> evaluation. This operator provides explicit functionality (especially
+> for arrays involving descriptors) that is analogous to the implicit push
+> of the base address of a structure prior to evaluation of a
+> DW_AT_data_member_location to access a data member of a structure. For
+> an example, see Appendix D.2 on page 304.
+
+
+### Section 2.5.2.1 Memory Locations
+
+Insert the following (adapted from Section 2.6.1.1.2):
+
+> A memory location represents the location of a piece or all of an
+> object or other entity in memory. On architectures that support
+> multiple address spaces, a memory location contains a component that
+> identifies the address space (which may be provided by the
+> `DW_OP_xderef` operation). A memory location is considered
+> _unbounded_, as the size of the location storage is only implied by
```

Comment:
IMHO, "location storage" is a confusing term, because the storage is not meant to store locations. Can we just say "storage"?

Reply:
Agreed, but I think "storage" is too generic and not obvious we're using a term for something specific. I've thought about other names for this, but the best I've come up with so far is "storage bank". What do you think?

Reply:
> Agreed, but I think "storage" is too generic and not obvious we're using a term for something specific. I've thought about other names for this, but the best I've come up with so far is "storage bank". What do you think?

I think it's better than "location storage".